qiskit/qpy/__init__.py

```diff
@@ -111,464 +138,473 @@
         uint8_t qiskit_major_version;
         uint8_t qiskit_minor_version;
         uint8_t qiskit_patch_version;
-        uint64_t num_circuits;
+        uint64_t num_circuits; // number of payload blocks
     }
 
-All values use network byte order [#f1]_ (big endian) for cross platform
-compatibility.
-
-The file header is immediately followed by the circuit payloads.
-Each individual circuit is composed of the following parts:
-
-``HEADER | METADATA | REGISTERS | CUSTOM_DEFINITIONS | INSTRUCTIONS``
-
-There is a circuit payload for each circuit (where the total number is dictated
-by ``num_circuits`` in the file header). There is no padding between the
-circuits in the data.
+.. note::
+    All values use network byte order [#f1]_ (big endian) for cross platform
+    compatibility.
 
-.. _qpy_version_7:
+``num_circuits`` defines the number of consecutive payload blocks that will be found in the file,
+with no padding in between data blocks.
 
-Version 7
-=========
+PAYLOAD_HEADER
+==============
 
-Version 7 adds support for :class:`.~Reference` instruction and serialization of
-a :class:`.~ScheduleBlock` program while keeping its reference to subroutines::
+In QPY v.5 and above, the payload header:
 
-    from qiskit import pulse
-    from qiskit import qpy
+.. code-block:: c
 
-    with pulse.build() as schedule:
-        pulse.reference("cr45p", "q0", "q1")
-        pulse.reference("x", "q0")
-        pulse.reference("cr45p", "q0", "q1")
+    struct {
+        char type;
+    }
 
-    with open('template_ecr.qpy', 'wb') as fd:
-        qpy.dump(schedule, fd)
+immediately follows the file header block to represent the program type stored in the file:
 
-The conventional :ref:`qpy_schedule_block` data model is preserved, but in
-version 7 it is immediately followed by an extra :ref:`qpy_mapping` utf8 bytes block
-representing the data of the referenced subroutines.
+- When ``type==c``, :class:`~.QuantumCircuit` payloads follow
+- When ``type==s``, :class:`~.ScheduleBlock` payloads follow
```

Comment:
Maybe link to the section too here. For circuit it's easy because it's right below. But for schedule block it's a lot of scrolling :)